### PR TITLE
feat(appeals): add change page application decision date

### PR DESCRIPTION
**Pull request by AlexMoulin**
## Issue ticket number and link
https://pins-ds.atlassian.net/browse/BOAT-1459
## Type of change 🧩

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Other (please explain in the description section above)

## Checklist before requesting a review

- [x] I have performed a self-review of my own code
- [x] I have double checked this work does not include any hardcoded secrets or passwords
- [x] I have made corresponding changes to the documentation
- [x] I have provided details on how I have tested my code
- [x] I have referenced the ticket number above
- [x] I have provided a description of the ticket
- [x] I have included unit tests to cover any testable code changes

### CODE REVIEW

File: appeals/web/src/server/appeals/appeal-details/appellant-case/__tests__/appellant-case.test.js [hunk 1]
```diff
@@ -135,6 +135,67 @@ describe('appellant-case', () => {
...
+
+			await request
+				.post(`${baseUrl}/${appealId}/appellant-case/application-decision-date/change`)
+				.send(validData);
+
+			const response = await request.get(`${baseUrl}/1${appellantCasePagePath}`);
+
+			const element = parseHtml(response.text);
+
+			expect(element.innerHTML).toMatchSnapshot();
```
Comment (beejjacobs) [inline appeals/web/src/server/appeals/appeal-details/appellant-case/__tests__/appellant-case.test.js:157]: Did you mean to have this too? Means this test includes a full-page snapshot. Applies to the one below as well.

File: appeals/web/src/server/appeals/appeal-details/application-decision-date/application-decision-date.controller.js [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+		return renderChangeApplicationHasDecisionDate(request, response);
+	}
+
+	try {
+		const radio = request.body['application-decision-radio'];
+		const { currentAppeal, apiClient } = request;
+
+		const { appealId, appellantCaseId } = currentAppeal;
+		if (radio === 'yes') {
+			request.session.radio = radio;
```
Comment (beejjacobs) [inline appeals/web/src/server/appeals/appeal-details/application-decision-date/application-decision-date.controller.js:69]: I don't see this being deleted anywhere, is it meant to be after its been used?

File: appeals/web/src/server/appeals/appeal-details/application-decision-date/application-decision-date.controller.js [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+		}
+	}
+};
+
+/**
+ * @param {import('@pins/express/types/express.js').Request} request
+ * @param {import('@pins/express/types/express.js').RenderedResponse<any, any, Number>} response
+ */
+export const postChangeApplicationSetDecisionDate = async (request, response) => {
+	console.log(request.body);
```
Comment (beejjacobs) [inline appeals/web/src/server/appeals/appeal-details/application-decision-date/application-decision-date.controller.js:141]: This been left in from dev perhaps?

File: appeals/web/src/server/appeals/appeal-details/application-decision-date/application-decision-date.controller.js [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+		const updatedApplicationDecisionDay = parseInt(session.applicationDecisionDate.day, 10);
+		const updatedApplicationDecisionMonth = parseInt(session.applicationDecisionDate.month, 10);
+		const updatedApplicationDecisionYear = parseInt(session.applicationDecisionDate.year, 10);
+
+		if (
+			Number.isNaN(updatedApplicationDecisionDay) ||
+			Number.isNaN(updatedApplicationDecisionMonth) ||
+			Number.isNaN(updatedApplicationDecisionYear)
+		) {
+			return response.status(500).render('app/500.njk');
```
Comment (beejjacobs) [inline appeals/web/src/server/appeals/appeal-details/application-decision-date/application-decision-date.controller.js:166]: I think 400 is appropriate here: Bad Request? Not sure if we have a page for it.

File: appeals/web/src/server/appeals/appeal-details/application-decision-date/application-decision-date.mapper.js [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+	const shortAppealReference = appealShortReference(appealData.appealReference);
+
+	let day = storedSessionData?.day ?? '';
+	let month = storedSessionData?.month ?? '';
+	let year = storedSessionData?.year ?? '';
+
+	if (appellantCaseData.applicationDecisionDate && !storedSessionData) {
+		const formattedApplicationDecisionDate = new Date(appellantCaseData.applicationDecisionDate);
+
+		day = formattedApplicationDecisionDate.getDate().toString();
```
Comment (beejjacobs) [inline appeals/web/src/server/appeals/appeal-details/application-decision-date/application-decision-date.mapper.js:80]: I don't know if we set the time on `applicationDecisionDate` to make this safe, but we might need to make sure we get the date in `Europe/London` as the server would be running in UTC.

**Review by beejjacobs — COMMENTED** [review]
Few comments/questions, but generally looks good!

File: appeals/web/src/server/appeals/appeal-details/application-decision-date/application-decision-date.controller.js (hunk 2 above)
Comment (AlexMoulin) [reply appeals/web/src/server/appeals/appeal-details/application-decision-date/application-decision-date.controller.js:69]: It was supposed to be in applicationDecisionDate

File: appeals/web/src/server/appeals/appeal-details/application-decision-date/application-decision-date.controller.js (hunk 4 above)
Comment (AlexMoulin) [reply appeals/web/src/server/appeals/appeal-details/application-decision-date/application-decision-date.controller.js:166]: There is no 400 page. I could create one if you think it's needed?

Comment (beejjacobs) [reply appeals/web/src/server/appeals/appeal-details/application-decision-date/application-decision-date.controller.js:166]: No that's OK, one to maybe visit post MVP. Validation should handle this anyway, so actually 500 is perhaps appropriate given that its unexpected!

File: appeals/web/src/server/appeals/appeal-details/application-decision-date/application-decision-date.controller.js (hunk 2 above)
Comment (AlexMoulin) [reply appeals/web/src/server/appeals/appeal-details/application-decision-date/application-decision-date.controller.js]: I ended up removing it anyway as it was causing issues with tests.

File: appeals/web/src/server/appeals/appeal-details/application-decision-date/application-decision-date.mapper.js (hunk 5 above)
Comment (AlexMoulin) [reply appeals/web/src/server/appeals/appeal-details/application-decision-date/application-decision-date.mapper.js]: Should be resolved now?

File: appeals/web/src/server/appeals/appeal-details/application-decision-date/application-decision-date.controller.js [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+ */
+export const postChangeApplicationHasDecisionDate = async (request, response) => {
+	if (request.errors) {
+		return renderChangeApplicationHasDecisionDate(request, response);
+	}
+
+	try {
+		const radio = request.body['application-decision-radio'];
+		const { currentAppeal, apiClient } = request;
+		console.log(radio);
```
Comment (beejjacobs) [inline appeals/web/src/server/appeals/appeal-details/application-decision-date/application-decision-date.controller.js:66]: Oops, one more to remove!

File: appeals/web/src/server/appeals/appeal-details/application-decision-date/application-decision-date.mapper.js (hunk 5 above)
Comment (james-hill-kc) [reply appeals/web/src/server/appeals/appeal-details/application-decision-date/application-decision-date.mapper.js:80]: Would `import { formatInTimeZone } from 'date-fns-tz';` be useful here?

**Review by james-hill-kc — APPROVED** [review]
LGTM, just added a suggestion to Ben's comment, unsure if it's helpful

File: appeals/web/src/server/appeals/appeal-details/application-decision-date/application-decision-date.mapper.js (hunk 5 above)
Comment (beejjacobs) [reply appeals/web/src/server/appeals/appeal-details/application-decision-date/application-decision-date.mapper.js:80]: Yes potentially. We agreed to leave this for now, as its just `YYYY-MM-DD` handling, it should be OK. Going to do a separate audit of date time handling to check.

**Review by beejjacobs — DISMISSED** [review]
LGTM!

**Review by james-hill-kc — APPROVED** [review]

**Review by kandcdev — APPROVED** [review]

**Pull request merged**